### PR TITLE
Wr/package uninstall

### PR DESCRIPTION
**Pull request by WillieRuemmele**
@W-10832862@

adds `uninstallPackage` method to uninstall a package, will optionally wait for uninstall to complete, by default won't poll
UTs

### CODE REVIEW

File: src/package/packageUninstall.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    default: {
+      const err = messages.getMessage('defaultErrorMessage', [id, uninstallRequest.Id]);
+      const errorQueryResult = await conn.tooling.query(
+        `"SELECT Message FROM PackageVersionUninstallRequestError WHERE ParentRequest.Id = '${id}' ORDER BY Message"`
+      );
+
+      const errors = [];
+      if (errorQueryResult.records.length) {
+        errors.push('\n=== Errors\n');
+        errorQueryResult?.records?.forEach((record) => {
```
Comment (peternhale) [inline src/package/packageUninstall.ts:40]: @WillieRuemmele are the safe navigation operators necessary here?

File: src/package/packageUninstall.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  });
+
+  if (wait.seconds === 0) {
+    return (await conn.tooling
+      .sobject('SubscriberPackageVersionUninstallRequest')
+      .retrieve(uninstallRequest.id)) as UninstallResult;
+  } else {
+    const pollingClient = await PollingClient.create({
+      poll: () => poll(uninstallRequest.id, conn),
+      frequency: Duration.seconds(5),
```
Comment (peternhale) [inline src/package/packageUninstall.ts:68]: @WillieRuemmele is an uninstall runtime pretty short, if so then 5 second interval would be ok, but if takes a long time then maybe this value is too short.

**Comment by peternhale** [comment]
@WillieRuemmele could you add the uninstall API call to this [nut test](https://github.com/forcedotcom/packaging/blob/135d4206ea130176053dd574c6b05ca9d78191bc/test/package/packageTest.nut.ts). There is an uninstall step that currently calls out to the sfdx command.

**Review by peternhale — CHANGES_REQUESTED** [review]

File: src/package/packageUninstall.ts (hunk 2 above)
Comment (WillieRuemmele) [reply src/package/packageUninstall.ts:68]: they were always pretty quick for me, it's also what toolbelt was using

**Review by mshanemc — COMMENTED** [review]
I have some meta-comments about how this is organized, not specific to this PR. Hopefully it's not too late to influence without making a mess.

1. For well-known types for the Tooling API, why not put them in jsforce instead of this library?
ex: https://github.com/jsforce/jsforce/blob/2.0/src/api/tooling.ts

2.  If that happened, you could do things the jsforce2 way ex:
```ts
const uninstallRequest = await conn.tooling
    .sobject<SubscriberPackageVersionUninstallRequest>('SubscriberPackageVersionUninstallRequest')
    .retrieve(id);
```
3. Let's avoid using enums, prefer string union types, so not:
```ts
export enum Package2VersionStatus {
    queued = 'Queued',
    inProgress = 'InProgress',
    ...
  }
```
instead
```
export type Package2VersionStatus = 'Queued' | 'InProgress' | ...
```

File: src/package/packageUninstall.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import * as os from 'os';
+import { Connection, Lifecycle, Messages, PollingClient, SfError, StatusResult } from '@salesforce/core';
+import { Duration } from '@salesforce/kit';
+import { PackagingSObjects } from '../interfaces';
+
+Messages.importMessagesDirectory(__dirname);
+const messages = Messages.loadMessages('@salesforce/packaging', 'messages');
```
Comment (shetzel) [inline src/package/packageUninstall.ts:13]: We don't necessarily have to change this now, but having 1 big messages markdown file is not what we should do.

File: src/package/packageUninstall.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+type UninstallResult = PackagingSObjects.SubscriberPackageVersionUninstallRequest;
+
+async function poll(id: string, conn: Connection): Promise<StatusResult> {
+  const uninstallRequest = await conn.tooling.sobject('SubscriberPackageVersionUninstallRequest').retrieve(id);
+
+  switch (uninstallRequest.Status) {
+    case 'Success': {
```
Comment (shetzel) [inline src/package/packageUninstall.ts:21]: Is this the actual casing returned by the server?  I'm guessing it is but that's not consistent with the install API, which returns status in all caps.  🙄

File: src/package/packageUninstall.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      if (errorQueryResult.records.length) {
+        errors.push('\n=== Errors\n');
+        errorQueryResult.records.forEach((record) => {
+          errors.push(`(${errors.length}) ${record.Message}${os.EOL}`);
+        });
+      }
+
+      throw new SfError(`${err}${errors.join(os.EOL)}`, 'UNINSTALL_ERROR', [
+        messages.getMessage('uninstallErrorAction'),
+      ]);
```
Comment (shetzel) [inline src/package/packageUninstall.ts:47]: You're right, this is a common function between install, install report, and uninstall.  We'll refactor later.

File: src/package/packageUninstall.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    }
+  }
+}
+
+export async function uninstallPackage(
+  id: string,
+  conn: Connection,
+  wait: Duration = Duration.seconds(0)
```
Comment (shetzel) [inline src/package/packageUninstall.ts:55]: Something to make consistent at some point is allowing both `number` and `Duration` for polling.  It makes for a friendlier API.

File: src/package/packageUninstall.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    return (await conn.tooling
+      .sobject('SubscriberPackageVersionUninstallRequest')
+      .retrieve(uninstallRequest.id)) as UninstallResult;
+  } else {
+    const pollingClient = await PollingClient.create({
+      poll: () => poll(uninstallRequest.id, conn),
+      frequency: Duration.seconds(5),
+      timeout: wait,
+    });
+    return pollingClient.subscribe();
```
Comment (shetzel) [inline src/package/packageUninstall.ts:71]: Without try/catch around this the polling client errors are thrown directly.  This is another place where we should be consistent.

**Review by shetzel — APPROVED** [review]

File: src/package/packageUninstall.ts (hunk 4 above)
Comment (WillieRuemmele) [reply src/package/packageUninstall.ts:21]: it is :(

**Review by peternhale — APPROVED** [review]

**Pull request merged**